velox/functions/sparksql/specialforms/DecimalRound.cpp

```diff
@@ -0,0 +1,217 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/DecimalRound.h"
+#include "velox/expression/ConstantExpr.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+template <typename R, typename A>
+class DecimalRoundFunction : public exec::VectorFunction {
+ public:
+  DecimalRoundFunction(
+      int32_t scale,
+      uint8_t inputPrecision,
+      uint8_t inputScale,
+      uint8_t resultPrecision,
+      uint8_t resultScale)
+      : scale_(scale),
+        inputPrecision_(inputPrecision),
+        inputScale_(inputScale),
+        resultPrecision_(resultPrecision),
+        resultScale_(resultScale) {
+    const auto [p, s] = DecimalRoundCallToSpecialForm::getResultPrecisionScale(
+        inputPrecision, inputScale, scale);
+    VELOX_USER_CHECK_EQ(
+        p,
+        resultPrecision,
+        "The result precision of decimal_round should be the same with expected.");
+    VELOX_USER_CHECK_EQ(
+        s,
+        resultScale,
+        "The result scale of decimal_round should be the same with expected.");
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_USER_CHECK(
+        args[0]->isConstantEncoding() || args[0]->isFlatEncoding(),
+        "Single-arg deterministic functions receive their only argument as flat or constant vector.");
+    context.ensureWritable(rows, resultType, result);
+    result->clearNulls(rows);
+    auto rawResults = result->asUnchecked<FlatVector<R>>()->mutableRawValues();
+    if (args[0]->isConstantEncoding()) {
+      // Fast path for constant vector.
+      applyConstant(rows, args[0], context, rawResults);
+    } else {
+      // Fast path for flat vector.
+      applyFlat(rows, args[0], context, rawResults);
+    }
+  }
+
+  bool supportsFlatNoNullsFastPath() const override {
+    return true;
+  }
+
+ private:
+  inline R applyRound(const A input) const {
+    if (scale_ >= 0) {
+      const auto rescaledValue = DecimalUtil::rescaleWithRoundUp<A, R>(
+          input, inputPrecision_, inputScale_, resultPrecision_, resultScale_);
+      VELOX_DCHECK(rescaledValue.has_value());
+      return rescaledValue.value();
+    } else {
+      const auto rescaleFactor =
+          DecimalUtil::kPowersOfTen[inputScale_ - scale_];
```

Comment:
Does this work when scale_ is a large negative number, i.e. -10'000?

Reply:
Thank you for catching this. It should be undefined behavior when accessing outside the bounds, but happens to work in this test. Fixed.

> scale: -10'000
rescaleFactor: 92076405166876042240236349289261904211

Reply:
Do we want undefined behavior? Wouldn't it be better to clamp 'scale' to a reasonable value or have an explicit check for values that are too large? Investigating wrong results or query failures caused by undefined behavior is never fun.

Reply:
No, we don't want that. I fixed this issue by using std::min to ensure the max index is not out side of bounds of the array.
> DecimalUtil::kPowersOfTen[std::min(
          inputScale_ - scale_, (int32_t)LongDecimalType::kMaxPrecision)];

> DecimalUtil::kPowersOfTen[std::min(
          -scale_, (int32_t)LongDecimalType::kMaxPrecision)]

Reply:
Thanks for the reminding.